app/code/Magento/CatalogRule/Model/Rule.php

```diff
@@ -517,9 +517,11 @@ protected function _invalidateCache()
     public function afterSave()
     {
         if ($this->isObjectNew()) {
-            $this->getMatchingProductIds();
-            if (!empty($this->_productIds) && is_array($this->_productIds)) {
-                $this->_getResource()->addCommitCallback([$this, 'reindex']);
+            if (!$this->_ruleProductProcessor->isIndexerScheduled()) {
+                $this->getMatchingProductIds();
```

Comment:
please, refactore this: use returned value for comparing instead of using internal property